eventsourcing/src/test/java/org/axonframework/eventsourcing/AnnotationBasedEventStateApplierTest.java

```diff
@@ -0,0 +1,343 @@
+            // given
+            var state = new TestState();
+            var event = domainEvent(0);
+
+            // when
+            state = eventStateApplier.apply(state, event, processingContext);
+
+            // then
+            assertEquals("null-0", state.handledSequences);
+        }
+
+        @Test
+        void resolvesSources() {
+            // given
+            var state = new TestState();
+            var event = domainEvent(0);
+
+            // when
+            state = eventStateApplier.apply(state, event, processingContext);
+
+            // then
+            assertEquals("null-id", state.handledSources);
+        }
+
+        @Test
+        void resolvesTimestamps() {
+            var timestamp = Instant.now();
+            GenericEventMessage.clock = Clock.fixed(timestamp, ZoneId.systemDefault());
+
+            // given
+            var state = new TestState();
+            var event = domainEvent(0);
+
+            // when
+            state = eventStateApplier.apply(state, event, processingContext);
+
+            // then
+            assertEquals("null-" + timestamp, state.handledTimestamps);
+        }
+    }
+
+    @Nested
+    class HandlerInvocationRules {
+
+        @Test
+        void doNotHandleNotDeclaredEventType() {
+            // given
+            var eventStateApplier = new AnnotationBasedEventStateApplier<>(HandlingJustStringState.class);
+            var state = new HandlingJustStringState();
+            var event = domainEvent(0);
+
+            // when
+            state = eventStateApplier.apply(state, event, processingContext);
+
+            // then
+            assertEquals(0, state.handledCount);
+        }
+
+        @Test
+        void invokesOnlyMostSpecificHandler() {
+            // given
+            var state = new TestState();
+            var event = domainEvent(0);
+
+            // when
+            state = eventStateApplier.apply(state, event, processingContext);
+
+            // then
+            assertEquals("null-0", state.handledPayloads);
+            assertFalse(state.objectHandlerInvoked);
+            assertEquals(1, state.handledCount);
+        }
+    }
+
+    // feel free to drop the support in case of some changed in EventHandling logic (like Empty stream always returned from event handling component)
+    @Nested
+    class RecordSupport {
+
+        private record RecordState(String handledPayloads) {
+
+            private static RecordState empty() {
+                return new RecordState("null");
+            }
+
+            @EventSourcingHandler
+            RecordState evolve(
+                    Integer payload
+            ) {
+                return new RecordState(handledPayloads + "-" + payload);
+            }
+        }
+
+        private static final EventStateApplier<RecordState> eventStateApplier = new AnnotationBasedEventStateApplier<>(
+                RecordState.class);
+
+        @Test
+        void doNotMutateStateIfRecord() {
```

Comment:
Test naming nit for clarity:
```suggestion
        void doNotMutateGivenStateIfRecord() {
```